Introduce custom logic into OnRecvPacket in PFM module to charge fee

RustNinja opened the pull request:
Introduce custom logic into OnRecvPacket in PFM module to charge fee

this code in custom pfm inside IbcMiddleware `OnRecvPacket` method introduced by me.
https://github.com/ComposableFi/composable-cosmos/blob/rustninja/pmf-middleware/custom/custompfm/keeper/keeper.go#L185-L216

all other code in method `OnRecvPacket` is taken from original version of `OnRecvPacket`.

RustNinja committed:
added custom pfm middleware to substitue logic on onRecvPacket

RustNinja committed:
import custom pfm into app.go (#517)

RustNinja committed:
use custom pfm instead of original router. all test passed.

RustNinja committed:
clone the logic from original version to custom pfm to introduce fees. (#518)

all test passed.

RustNinja committed:
Introduce custom logic into OnRecvPacket in PFM module to charge fee to prevent a spams (#519)

All test passed.

RustNinja committed:
remove custom logic in ibc transfer keeper.

blasrodri commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,311 @@
+	}
+}
+
+func (im IBCMiddleware) OnRecvPacket(
```
don't you need to implement all of the methods?

RustNinja replied:
no, all other methods are implemented because i include into mine IbcMidleware type the original IbcMiddleware that bring implementation of others methods as inherited from original
I just override the only one method to inject my custom fee logic

rjonczy reviewed (changes requested):
Why do you target `develop2` ?

target: `main` (current mainnet) or `testnet` (current testnet)

vmarkushin commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,311 @@
+		// unwind denom
+		unwoundDenom := denom[len(counterpartyPrefix):]
+		denomTrace := transfertypes.ParseDenomTrace(unwoundDenom)
+		if denomTrace.Path == "" {
+			// denom is now unwound back to native denom
+			return unwoundDenom
+		}
+		// denom is still IBC denom
+		return denomTrace.IBCDenom()
```
Can't it be replaced with `return denomTrace.IBCDenom()`?

RustNinja replied:
this is not my code. this from original version of IBCMiddleware. so instead of fork all module, i just copy pasted the current function and inject extra code with fee logic.

vmarkushin commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,311 @@
+			im.keeper1.Logger(ctx).Error("packetForwardMiddleware error marshaling next as JSON",
+				"error", err,
+			)
+			// return errorsmod.Wrapf(sdkerrors.ErrJSONMarshal, err.Error())
```
Why we don't return an error here?

RustNinja replied:
This error will be handled by original pfm in case if memo was incorrect.

vmarkushin commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,311 @@
+		return newErrorAcknowledgement(err)
+	}
+
+	// override the receiver so that senders cannot move funds through arbitrary addresses.
```
Why he shouldn't be able to do that?

RustNinja replied:
this is not my code. this from original version of IBCMiddleware. so instead of fork all module, i just copy pasted the current function and inject extra code with fee logic.

vmarkushin commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,311 @@
+			RevisionNumber: 0,
+			RevisionHeight: 0,
+		},
+		uint64(ctx.BlockTime().UnixNano())+uint64(timeout.Nanoseconds()),
```
Is `timeout.Nanoseconds()` offset or absolute?

RustNinja replied:
this is not my code. this from original version of IBCMiddleware. so instead of fork all module, i just copy pasted the current function and inject extra code with fee logic.

vmarkushin replied:
Could you send the link to it?

blasrodri commented:
Please include an interchain test here

RustNinja committed:
fix build

RustNinja commented:
> Why do you target `develop2` ?
> 
> target: `main` (current mainnet) or `testnet` (current testnet)

this is devnet. first devnet. then will merge into testnet. and will go to testnet.

vmarkushin reviewed (commented):
Please highlight the changes to the middleware in the code so it's clear where it's the original middleware and where it's our changes

RustNinja commented:
> Please highlight the changes to the middleware in the code so it's clear where it's the original middleware and where it's our changes

at the description

i duplicate here too
https://github.com/ComposableFi/composable-cosmos/blob/rustninja/pmf-middleware/custom/custompfm/keeper/keeper.go#L185-L216

vmarkushin commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,311 @@
+		retries = im.retriesOnTimeout1
+	}
+
+	// im.ibcfeekeeper.Transfer()
```
Remove such commented code everywhere, otherwise leave a comment explaining why it's needed

RustNinja replied:
Done.

vmarkushin commented on x/ibctransfermiddleware/keeper/keeper.go:
```diff
@@ -104,3 +109,111 @@ func (k Keeper) GetChannelFeeAddress(ctx sdk.Context, targetChannelID string) st
+	Receiver sdk.AccAddress
+}
+
+func (k Keeper) ChargeFee(ctx sdk.Context, msg *ibctypes.MsgTransfer) (*BridgeFee, error) {
```
Why `ChargeFee` has something to do with timeouts?

RustNinja replied:
I will rename function
thanks

RustNinja replied:
Renamed.

vmarkushin commented on x/ibctransfermiddleware/keeper/keeper.go:
```diff
@@ -104,3 +109,111 @@ func (k Keeper) GetChannelFeeAddress(ctx sdk.Context, targetChannelID string) st
+				return nil, fmt.Errorf("token not allowed to be transferred in this channel")
+			}
+
+			minFee := coin.MinFee.Amount
```
Does this function takes all the fees or only some specific ones?

RustNinja replied:
function
calculate fee based on config as percentage fee + min fee for some specific demon

vmarkushin replied:
Is this fee something different? Can't we use only it?

https://github.com/ComposableFi/composable-cosmos/blob/rustninja/pmf-middleware/custom/custompfm/keeper/keeper.go#L167

RustNinja replied:
No. this fee is the logic from original pfm.
it is different.

our requirements is to add extra field depends on channel and denom. where each fee contains 3 params
account where to send this fee, percentage from amount + min amount equal to 20$.

RustNinja committed:
rename fn and remove commented code.

christianvari commented on x/ibctransfermiddleware/keeper/keeper.go:
```diff
@@ -104,3 +109,111 @@ func (k Keeper) GetChannelFeeAddress(ctx sdk.Context, targetChannelID string) st
+
+			minFee := coin.MinFee.Amount
+			priority := GetPriority(msg.Memo)
+			if priority != nil {
+				p := findPriority(coin.TxPriorityFee, *priority)
+				if p != nil && coin.MinFee.Denom == p.PriorityFee.Denom {
+					minFee = minFee.Add(p.PriorityFee.Amount)
+				}
+			}
```
If the sender specifies in the `memo` a `priority` coin that is not in the `TxPriorityFee` or with a `Denom` different from the `MinFee` one, it will be not charged and no error raised. Is this intentional?

RustNinja replied:
Yes. So FE is responsible to set a proper priority (high, medium, low) depends on config in storage for this denom.

christianvari commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,309 @@
+		return newErrorAcknowledgement(fmt.Errorf("error charging fee: %w", err))
+	}
+	if result != nil {
+		if result.Fee.Amount.LT(token.Amount) {
+			token = token.SubAmount(result.Fee.Amount)
+		} else {
```
`result.Fee` is calculated on `PacketCoin` which is `token` deducted by the `pfm` fees. 
However since here we deduct `token` instead of `PacketCoin` and then pass it to `ForwardTransferPacket` which applies `pfm` fees to it, we will get inaccurate calculations for non-fixed fees

RustNinja replied:
It depends how to specify the order of charging a fee.
charge bridge fee when pfm already was charged or opposite.

But now problem. i will remove extra code where i intentionally first determinate the pfm fee and use new amount to charge extra bridge fee for specific channel and denom. 

I got your point. Accepted your request.

christianvari commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,309 @@
+		logger.Error("packetForwardMiddleware OnRecvPacket error charging fee", "error", err)
+		return newErrorAcknowledgement(fmt.Errorf("error charging fee: %w", err))
+	}
+	if result != nil {
+		if result.Fee.Amount.LT(token.Amount) {
+			token = token.SubAmount(result.Fee.Amount)
+		} else {
+			send_err := im.bank.SendCoins(ctx, result.Sender, result.Receiver, sdk.NewCoins(result.Fee))
```
In case `result.Fee.Amount.LT(token.Amount)` we are deducting the `Fee` from `token`, however, fees are not sent to the `feeAddress` and will remain in the receiver account

RustNinja replied:
It is really good point.
Thanks.
Updated the logic.

christianvari commented on custom/custompfm/keeper/keeper.go:
```diff
@@ -0,0 +1,309 @@
+	}
+	if result != nil {
+		if result.Fee.Amount.LT(token.Amount) {
+			token = token.SubAmount(result.Fee.Amount)
+		} else {
+			send_err := im.bank.SendCoins(ctx, result.Sender, result.Receiver, sdk.NewCoins(result.Fee))
+			if send_err != nil {
+				logger.Error("packetForwardMiddleware OnRecvPacket error sending fee", "error", send_err)
+				return newErrorAcknowledgement(fmt.Errorf("error charging fee: %w", send_err))
+			}
+			ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
```
I'm not sure if the `SendCoins` here would work correctly. Funds at this point have already been received by the `overrideReceiver` so they should be transferred from this account probably.
https://github.com/ComposableFi/composable-cosmos/blob/47c2ed65e657e46e2b005de22846f766262484de/custom/custompfm/keeper/keeper.go#L110-L124

RustNinja replied:
@christianvari 
thank you a lot for warning.

you are correct. assets already sent to override receiver address. so fee should be sent from that account.
And this is exactly what our custom logic is doing.

Look.

```
tr := transfertypes.NewMsgTransfer(
		metadata.Port,
		metadata.Channel,
		token,
		overrideReceiver,
		metadata.Receiver,
		clienttypes.Height{
			RevisionNumber: 0,
			RevisionHeight: 0,
		},
		uint64(ctx.BlockTime().UnixNano())+uint64(timeout.Nanoseconds()),
		memo,
	)

	result, err := im.ibcfeekeeper.GetBridgeFeeBasedOnConfigForChannelAndDenom(ctx, tr)
```

I created `tr` variable passing the exact `overrideReceiver` as a `sender` 
So the `result` from `GetBridgeFeeBasedOnConfigForChannelAndDenom` contains a field `sender` that equal to `overrideReceiver`.
So all good. 
Thank for pointing and double check.

RustNinja committed:
requested changes by auditor. (#524)

- remove calculation of pfm fee before request calculation of extra
birdge fee for channel and denom if exists
- fix a bug when send to fee account only one case when fee is less then
token amount. so it was a bug. fixed. now send coin always when fee is
less then token amount or equal.

hoank101 committed:
fix: some issues when upgrade to sdk50 (#531)

Co-authored-by: dzmitry-lahoda <dzmitry@lahoda.pro>
Co-authored-by: kienn6034 <kien@notional.ventures>
Co-authored-by: rustdev <placex.com@gmail.com>
Co-authored-by: rust.dev <102041955+RustNinja@users.noreply.github.com>
Co-authored-by: kkast <kkastsevich@gmail.com>
Co-authored-by: Kanstantsin Kastsevich <kkast@users.noreply.github.com>
Co-authored-by: rjonczy <robert.jonczy@gmail.com>
Co-authored-by: tungle <anhletung@notional.ventures>

hoank101 committed:
Merge branch 'develop2' of https://github.com/ComposableFi/composable-cosmos into rustninja/pmf-middleware

# Conflicts:
#	app/keepers/keepers.go
#	x/ibctransfermiddleware/keeper/keeper.go

hoank101 committed:
fix: some issues when update to sdk50
